src/com/owncloud/android/files/services/FileUploadService.java

```diff
@@ -0,0 +1,1359 @@
+     *         uploads.
+     */
+    private static boolean chunkedUploadIsSupported(OwnCloudVersion version) {
+        return (version != null && version.compareTo(OwnCloudVersion.owncloud_v4_5) >= 0);
+    }
+
+    /**
+     * Service initialization
+     */
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        Log_OC.d(TAG, "mPendingUploads size:" + mPendingUploads.size() + " - onCreate");
+        mNotificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
+        mBinder = new FileUploaderBinder();
+        mDb = UploadDbHandler.getInstance(this.getBaseContext());
+        mDb.recreateDb(); //for testing only
+        
+        //when this service starts there is no upload in progress. if db says so, app probably crashed before.
+        mDb.setAllCurrentToUploadLater();
+        
+        HandlerThread thread = new HandlerThread("FileUploadService-Requester");
+        thread.start();
+        
+        mServiceLooper = thread.getLooper();
+        mServiceHandler = new ServiceHandler(mServiceLooper);
+
+        uploadExecutor = Executors.newFixedThreadPool(1);
+
+        Log_OC.d(TAG, "FileUploadService.retry() called by onCreate()");
+        FileUploadService.retry(getApplicationContext());
+    }
+    
+    @Override
+    public void onDestroy() {
+        Log_OC.d(TAG, "mPendingUploads size:" + mPendingUploads.size() + " - onDestroy");
+        mServiceLooper.quit();
+        uploadExecutor.shutdown();
+    }
+
+    private final class ServiceHandler extends Handler {
+        public ServiceHandler(Looper looper) {
+            super(looper);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
+            onHandleIntent((Intent) msg.obj, (int)msg.arg1);
+        }
+    }
+
+    @Override
+    public void onStart(Intent intent, int intentStartId) {
+        Message msg = mServiceHandler.obtainMessage();
+        msg.arg1 = intentStartId;
+        msg.obj = intent;
+        mServiceHandler.sendMessage(msg);
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        onStart(intent, startId);
+        return START_STICKY; // if service is killed by OS before calling
+                             // stopSelf(), tell OS to restart service with
+                             // null-intent. 
+    }
+
+    /**
+     * The IntentService calls this method from the default worker thread with
+     * the intent that started the service. When this method returns,
+     * IntentService stops the service, as appropriate.
+     * 
+     * Note: We use an IntentService here. It does not provide simultaneous
+     * requests, but instead internally queues them and gets them to this
+     * onHandleIntent method one after another. This makes implementation less
+     * error-prone but prevents files to be added to list while another upload
+     * is active. If everything else works fine, fixing this should be a TODO.
```

Comment:
As far as I know, ```FileService``` was both queueing upload requests, running a single upload according to the queue, and stopping the service when no upload was pending, all of it correctly. I would really appreciate to know of any condition where any of this was failing. This is not irony, I would really like to know about it. ```IntentService``` is a commodity, really useful in many situations, but I think that changing all this code is totally unnecessary for the initial target: making uploads more reliable.

I understand that the code receiving upload requests in ```FileService``` was ugly as hell, and sure there is some other easier way to do it, but it was done the way it was due to a powerfull reason: keeping the associative array ```mPendingUploads``` so that it was possible to cancel any of the requested uploads, no matter if already in progress or waiting. I noticed you kept this array, but I still need to read the changes on the cancellation process. 

And probably will need to re-read this code (```onCreate```, ```onStartCommand```, ```onStart```, ```onHandleIntent```) a couple of times to really understand how it works.